File: batch_base64.rb
# batch_base64.rb — Ruby версия

require 'base64'
require 'find'
require 'json'
require 'time'
require 'thread'

class BatchBase64
  attr_reader :input_dir, :output_dir, :threads, :decode, :recursive, :report
  attr_accessor :processed, :errors, :failed, :start_time, :end_time

  def initialize(input_dir, output_dir = nil, threads = 4, decode = false, recursive = true, report = true)
    @input_dir = input_dir
    @output_dir = output_dir || File.join(input_dir, 'base64_output')
    @threads = threads
    @decode = decode
    @recursive = recursive
    @report = report
    @processed = 0
    @errors = 0
    @failed = []
    @mutex = Mutex.new
  end

  def get_files
    files = []
    if @recursive
      Find.find(@input_dir) do |path|
        files << path if File.file?(path)
      end
    else
      Dir.entries(@input_dir).each do |entry|
        path = File.join(@input_dir, entry)
        files << path if File.file?(path)
      end
    end
    files
  end

  def process_file(file_path)
    rel_path = file_path.sub(/^#{Regexp.escape(@input_dir)}\//, '')
    if @decode
      output_path = File.join(@output_dir, rel_path.gsub(/\.b64$/, ''))
    else
      output_path = File.join(@output_dir, rel_path + '.b64')
    end

    FileUtils.mkdir_p(File.dirname(output_path))

    begin
      if @decode
        content = File.read(file_path)
        decoded = Base64.decode64(content)
        File.write(output_path, decoded, mode: 'wb')
      else
        data = File.read(file_path, mode: 'rb')
        encoded = Base64.encode64(data).strip
        File.write(output_path, encoded)
      end
      @mutex.synchronize { @processed += 1 }
      true
    rescue => e
      @mutex.synchronize do
        @errors += 1
        @failed << { path: rel_path, error: e.message }
      end
      false
    end
  end

  def run
    puts "\e[36m🔐 Base64 Batch Encoder (Ruby)\e[0m"
    puts "📁 Папка: #{@input_dir}"

    files = get_files
    puts "📂 Найдено #{files.size} файлов"
    puts "⚡ Параллельная обработка (#{@threads} потоков)...\n"

    @start_time = Time.now

    queue = Queue.new
    files.each { |f| queue << f }

    workers = @threads.times.map do
      Thread.new do
        while !queue.empty? && (file = queue.pop(true) rescue nil)
          process_file(file)
        end
      end
    end
    workers.each(&:join)

    @end_time = Time.now
    elapsed = @end_time - @start_time

    puts
    puts "\e[32m✅ Обработано: #{@processed} файлов\e[0m"
    puts "\e[33m⚠️ Ошибок: #{@errors}\e[0m"
    puts "\e[36m⏱️ Время: #{elapsed.round(2)} сек\e[0m"
    puts "\e[32m💾 Сохранено: #{@output_dir}\e[0m"

    save_report(elapsed) if @report
  end

  def save_report(elapsed)
    report = {
      input_dir: @input_dir,
      output_dir: @output_dir,
      total_files: @processed + @errors,
      processed: @processed,
      errors: @errors,
      failed: @failed,
      start_time: @start_time.iso8601,
      end_time: @end_time.iso8601,
      elapsed_seconds: elapsed,
      decode_mode: @decode
    }
    report_path = File.join(@output_dir, 'report.json')
    File.write(report_path, JSON.pretty_generate(report))
    puts "\e[32m📊 Отчёт: #{report_path}\e[0m"
  end
end

def main
  input_dir = nil
  output_dir = nil
  threads = 4
  decode = false
  recursive = true
  report = true

  args = ARGV
  i = 0
  while i < args.size
    case args[i]
    when '--output', '-o'
      output_dir = args[i+1]; i += 2
    when '--threads', '-t'
      threads = args[i+1].to_i; i += 2
    when '--decode', '-d'
      decode = true; i += 1
    when '--no-recursive'
      recursive = false; i += 1
    when '--no-report'
      report = false; i += 1
    else
      input_dir = args[i] if input_dir.nil?
      i += 1
    end
  end

  unless input_dir
    puts "Usage: ruby batch_base64.rb <input_dir> [--output dir] [--threads N] [--decode]"
    exit 1
  end

  unless Dir.exist?(input_dir)
    puts "\e[31m❌ Папка не найдена: #{input_dir}\e[0m"
    exit 1
  end

  batch = BatchBase64.new(input_dir, output_dir, threads, decode, recursive, report)
  batch.run
end

main if __FILE__ == $0

File: batch_base64.rs
// batch_base64.rs — Rust версия

use rayon::prelude::*;
use serde_json::json;
use std::fs;
use std::path::{Path, PathBuf};
use std::time::Instant;
use walkdir::WalkDir;
use base64::{Engine as _, engine::general_purpose};

struct BatchBase64 {
    input_dir: PathBuf,
    output_dir: PathBuf,
    threads: usize,
    decode: bool,
    recursive: bool,
    report: bool,
    processed: usize,
    errors: usize,
    failed: Vec<(String, String)>,
}

impl BatchBase64 {
    fn new(input_dir: &str, output_dir: Option<&str>, threads: usize, decode: bool, recursive: bool, report: bool) -> Self {
        let input_path = PathBuf::from(input_dir);
        let output_path = output_dir.map_or_else(
            || input_path.join("base64_output"),
            |d| PathBuf::from(d)
        );
        BatchBase64 {
            input_dir: input_path,
            output_dir: output_path,
            threads,
            decode,
            recursive,
            report,
            processed: 0,
            errors: 0,
            failed: Vec::new(),
        }
    }

    fn get_files(&self) -> Vec<PathBuf> {
        let mut files = Vec::new();
        if self.recursive {
            for entry in WalkDir::new(&self.input_dir) {
                if let Ok(entry) = entry {
                    if entry.file_type().is_file() {
                        files.push(entry.path().to_path_buf());
                    }
                }
            }
        } else {
            if let Ok(entries) = fs::read_dir(&self.input_dir) {
                for entry in entries.flatten() {
                    if entry.file_type().map_or(false, |ft| ft.is_file()) {
                        files.push(entry.path());
                    }
                }
            }
        }
        files
    }

    fn process_file(&mut self, file_path: &Path) -> Result<(), String> {
        let rel_path = file_path.strip_prefix(&self.input_dir).unwrap_or(file_path);
        let rel_str = rel_path.to_str().unwrap_or("");

        let output_path = if self.decode {
            let name = if rel_str.ends_with(".b64") {
                &rel_str[..rel_str.len()-4]
            } else {
                rel_str
            };
            self.output_dir.join(name)
        } else {
            self.output_dir.join(format!("{}.b64", rel_str))
        };

        if let Some(parent) = output_path.parent() {
            fs::create_dir_all(parent).map_err(|e| format!("Ошибка создания папки: {}", e))?;
        }

        if self.decode {
            let content = fs::read_to_string(file_path).map_err(|e| format!("Ошибка чтения: {}", e))?;
            let decoded = general_purpose::STANDARD.decode(content.trim()).map_err(|e| format!("Ошибка декодирования: {}", e))?;
            fs::write(output_path, decoded).map_err(|e| format!("Ошибка записи: {}", e))?;
        } else {
            let data = fs::read(file_path).map_err(|e| format!("Ошибка чтения: {}", e))?;
            let encoded = general_purpose::STANDARD.encode(&data);
            fs::write(output_path, encoded).map_err(|e| format!("Ошибка записи: {}", e))?;
        }

        Ok(())
    }

    fn run(&mut self) {
        println!("\x1b[36m🔐 Base64 Batch Encoder (Rust)\x1b[0m");
        println!("📁 Папка: {}", self.input_dir.display());

        let files = self.get_files();
        println!("📂 Найдено {} файлов", files.len());
        println!("⚡ Параллельная обработка ({} потоков)...\n", self.threads);

        let start = Instant::now();

        // Используем rayon для параллельной обработки
        let results: Vec<(PathBuf, Result<(), String>)> = files
            .par_iter()
            .map(|file| {
                let mut result = (file.clone(), Ok(()));
                if let Err(e) = self.process_file(file) {
                    result.1 = Err(e);
                }
                result
            })
            .collect();

        for (file, res) in results {
            match res {
                Ok(()) => { self.processed += 1; }
                Err(e) => {
                    self.errors += 1;
                    let rel = file.strip_prefix(&self.input_dir).unwrap_or(&file);
                    self.failed.push((rel.to_str().unwrap_or("").to_string(), e));
                }
            }
        }

        let elapsed = start.elapsed().as_secs_f64();

        println!();
        println!("\x1b[32m✅ Обработано: {} файлов\x1b[0m", self.processed);
        println!("\x1b[33m⚠️ Ошибок: {}\x1b[0m", self.errors);
        println!("\x1b[36m⏱️ Время: {:.2} сек\x1b[0m", elapsed);
        println!("\x1b[32m💾 Сохранено: {}\x1b[0m", self.output_dir.display());

        if self.report {
            self.save_report(elapsed);
        }
    }

    fn save_report(&self, elapsed: f64) {
        let report = json!({
            "input_dir": self.input_dir,
            "output_dir": self.output_dir,
            "total_files": self.processed + self.errors,
            "processed": self.processed,
            "errors": self.errors,
            "failed": self.failed.iter().map(|(p, e)| json!({"path": p, "error": e})).collect::<Vec<_>>(),
            "start_time": chrono::Local::now().to_rfc3339(),
            "end_time": chrono::Local::now().to_rfc3339(),
            "elapsed_seconds": elapsed,
            "decode_mode": self.decode
        });
        let report_path = self.output_dir.join("report.json");
        let json_str = serde_json::to_string_pretty(&report).unwrap();
        fs::write(report_path, json_str).unwrap();
        println!("\x1b[32m📊 Отчёт: {}/report.json\x1b[0m", self.output_dir.display());
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let mut input_dir = None;
    let mut output_dir = None;
    let mut threads = rayon::current_num_threads();
    let mut decode = false;
    let mut recursive = true;
    let mut report = true;

    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--output" | "-o" => { output_dir = Some(args[i+1].clone()); i += 2; }
            "--threads" | "-t" => { threads = args[i+1].parse().unwrap_or(threads); i += 2; }
            "--decode" | "-d" => { decode = true; i += 1; }
            "--no-recursive" => { recursive = false; i += 1; }
            "--no-report" => { report = false; i += 1; }
            _ => {
                if input_dir.is_none() && !args[i].starts_with("-") {
                    input_dir = Some(args[i].clone());
                }
                i += 1;
            }
        }
    }

    if input_dir.is_none() {
        println!("Usage: cargo run -- <input_dir> [--output dir] [--threads N] [--decode]");
        std::process::exit(1);
    }

    let input = input_dir.unwrap();
    if !Path::new(&input).exists() {
        println!("\x1b[31m❌ Папка не найдена: {}\x1b[0m", input);
        std::process::exit(1);
    }

    let mut batch = BatchBase64::new(&input, output_dir.as_deref(), threads, decode, recursive, report);
    batch.run();
}
